pkg/cloud/services/compute/instance.go

```diff
@@ -568,36 +567,38 @@ func (s *Service) AssociateFloatingIP(instanceID, floatingIP string) error {
 }
 
 func (s *Service) InstanceDelete(machine *clusterv1.Machine, openStackMachine *infrav1.OpenStackMachine) error {
-	if machine.Spec.ProviderID == nil {
+	if openStackMachine == nil {
 		// nothing to do
 		return nil
 	}
 
-	parsed, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
+	instance, err := s.InstanceExists(openStackMachine.Name)
 	if err != nil {
 		return err
 	}
-	if err = deleteInstance(s, parsed.ID()); err != nil {
-		record.Warnf(openStackMachine, "FailedDeleteServer", "Failed to deleted server %s with id %s: %v", openStackMachine.Name, parsed.ID(), err)
-		return err
-	}
+	if instance != nil && instance.ID != "" {
```

Comment:
I think we should use a guard clause here:
```go
if instance == nil || instance.ID == "" {
  return nil
}
```

Reply:
Fixed.